vendor/k8s.io/kubernetes/pkg/controller/certificates/group_approver.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificates
+
+import (
+	"reflect"
+	"strings"
+
+	"k8s.io/kubernetes/pkg/apis/certificates"
+	unversionedcertificates "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/certificates/unversioned"
+	utilcertificates "k8s.io/kubernetes/pkg/util/certificates"
+
+	"github.com/golang/glog"
+)
+
+type groupApprover struct {
+	client                            unversionedcertificates.CertificateSigningRequestInterface
+	approveAllKubeletCSRsForGroup     string
+	approveKubeletServingCSRsForGroup string
+}
+
+// NewGroupApprover creates an approver that
+func NewGroupApprover(client unversionedcertificates.CertificateSigningRequestInterface, approveAllKubeletCSRsForGroup, approveKubeletServingCSRsForGroup string) CertificateAutoApprover {
+	return &groupApprover{
+		client: client,
+		approveAllKubeletCSRsForGroup:     approveAllKubeletCSRsForGroup,
+		approveKubeletServingCSRsForGroup: approveKubeletServingCSRsForGroup,
+	}
+}
+
+func (a *groupApprover) AutoApprove(csr *certificates.CertificateSigningRequest) (*certificates.CertificateSigningRequest, error) {
+	// short-circuit if we're already approved or denied
+	if approved, denied := getCertApprovalCondition(&csr.Status); approved || denied {
+		return csr, nil
+	}
+
+	var signClient, signServer bool
+	for _, g := range csr.Spec.Groups {
+		if len(a.approveAllKubeletCSRsForGroup) > 0 && g == a.approveAllKubeletCSRsForGroup {
+			signClient = true
+			break
+		}
+		if len(a.approveKubeletServingCSRsForGroup) > 0 && g == a.approveKubeletServingCSRsForGroup {
```

Comment:
If you sign serving certificates without confirming that my identity actually owns the hostname that you're signing for, then I can exploit the API to gain a serving cert for another node which in combination with an owned node network will allow me to steal any used secret in the system even if we close the API authorizer gaps.

I haven't dug into the pull, but if you allowed me to get signed cert for the address of the master, I could also steal any node credentials I wanted.